Tools/TestWebKitAPI/Tests/ios/DataListTextSuggestionTests.mm

```diff
@@ -0,0 +1,174 @@
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#import "config.h"
+
+#if ENABLE(DATALIST_ELEMENT) && PLATFORM(IOS_FAMILY)
+
+#import "PlatformUtilities.h"
+#import "TestInputDelegate.h"
+#import "TestWKWebView.h"
+#import "UIKitSPIForTesting.h"
+#import <WebKit/WKWebViewPrivate.h>
+#import <pal/spi/ios/BrowserEngineKitSPI.h>
+
+@interface DataListTextSuggestionInputDelegate : NSObject <UITextInputSuggestionDelegate
+#if USE(BROWSERENGINEKIT)
+    , BETextInputDelegate
+#endif
+>
+@end
+
+@implementation DataListTextSuggestionInputDelegate {
+    RetainPtr<NSArray> _suggestions;
+}
+
+- (NSArray *)suggestions
+{
+    return _suggestions.get();
+}
+
+- (void)setSuggestions:(NSArray *)suggestions
+{
+    _suggestions = suggestions;
+}
+
+- (void)selectionDidChange:(id<UITextInput>)textInput
+{
+}
+
+- (void)selectionWillChange:(id<UITextInput>)textInput
+{
+}
+
+- (void)textDidChange:(id<UITextInput>)textInput
+{
+}
+
+- (void)textWillChange:(id<UITextInput>)textInput
+{
+}
+
+- (void)handleWebViewCredentialsSaveForWebsiteURL:(NSURL *)websiteURL user:(NSString *)user password:(NSString *)password passwordIsAutoGenerated:(BOOL)passwordIsAutoGenerated
+{
+}
+
+#if USE(BROWSERENGINEKIT)
+
+- (void)invalidateTextEntryContextForTextInput:(id<BETextInput>)textInput
+{
+}
+
+- (void)selectionDidChangeForTextInput:(id<BETextInput>)textInput {
+
+}
+
+- (void)selectionWillChangeForTextInput:(id<BETextInput>)textInput {
+
+}
+
+- (BOOL)shouldDeferEventHandlingToSystemForTextInput:(id<BETextInput>)textInput context:(BEKeyEntryContext *)keyEventContext
+{
+    return NO;
+}
+
+- (void)textInput:(id<BETextInput>)textInput deferReplaceTextActionToSystem:(id)sender
+{
+}
+
+- (void)textInput:(id<BETextInput>)textInput setCandidateSuggestions:(NSArray *)suggestions
+{
+    RetainPtr uikitSuggestions = [NSMutableArray arrayWithCapacity:suggestions.count];
+    for (BETextSuggestion *beTextSuggestion in suggestions)
+        [uikitSuggestions addObject:beTextSuggestion._uikitTextSuggestion];
+
+    [self setSuggestions:uikitSuggestions.get()];
```

Comment:
Does `createNSArray` accept an `NSArray *` to do the mapping?
```suggestion
    [self setSuggestions:createNSArray(suggestions, [](BETextSuggestion *beTextSuggestion) {
        return beTextSuggestion._uikitTextSuggestion;
    }];
```

Reply:
If not, then what you have is good

Reply:
To my knowledge, it does not.